python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+            if "RAY_ADDRESS" in os.environ:
+                address = os.environ["RAY_ADDRESS"]
+                print("Connecting to ray cluster at address='{}'".format(
+                    address))
+                ray.init(address=address)
+            # Local mode.
+            else:
+                print("Starting local ray cluster")
+                ray.init(num_cpus=processes)
+
+        ray_cpus = int(ray.state.cluster_resources()["CPU"])
+        if processes is None:
+            processes = ray_cpus
+        elif ray_cpus < processes:
+            raise ValueError("Tried to start a pool with {} processes on an "
+                             "existing ray cluster, but there are only {} "
+                             "CPUs in the ray cluster.".format(
+                                 processes, ray_cpus))
+
+        return processes
+
+    def _start_actor_pool(self, processes):
+        self._actor_pool = [self._new_actor_entry() for _ in range(processes)]
+        ray.get([actor.ping.remote() for actor, _ in self._actor_pool])
+
+    def _wait_for_stopping_actors(self, timeout=None):
+        if len(self._actor_deletion_ids) == 0:
+            return
+        if timeout is not None:
+            timeout = float(timeout)
+
+        _, deleting = ray.wait(
+            self._actor_deletion_ids,
+            num_returns=len(self._actor_deletion_ids),
+            timeout=timeout)
+        self._actor_deletion_ids = deleting
+
+    def _stop_actor(self, actor):
+        # Check and clean up any outstanding IDs corresponding to deletions.
+        self._wait_for_stopping_actors(timeout=0.0)
+        # The deletion task will block until the actor has finished executing
+        # all pending tasks.
+        self._actor_deletion_ids.append(actor.__ray_terminate__.remote())
+
+    def _new_actor_entry(self):
+        # TODO(edoakes): The initializer function can't currently be used to
+        # modify the global namespace (e.g., import packages or set globals)
+        # due to a limitation in cloudpickle.
+        return (PoolActor.remote(self._initializer, self._initargs), 0)
+
+    # Batch should be a list of tuples: (args, kwargs).
+    def _run_batch(self, actor_index, func, batch):
+        actor, count = self._actor_pool[actor_index]
+        object_id = actor.run_batch.remote(func, batch)
+        count += 1
+        if count == self._maxtasksperchild:
+            self._stop_actor(actor)
+            actor, count = self._new_actor_entry()
+        self._actor_pool[actor_index] = (actor, count)
+        return object_id
+
+    def apply(self, func, args=None, kwargs=None):
+        return self.apply_async(func, args, kwargs).get()
+
+    def apply_async(self,
+                    func,
+                    args=None,
+                    kwargs=None,
+                    callback=None,
+                    error_callback=None):
+        self._check_running()
+        random_actor_index = random.randrange(len(self._actor_pool))
```

Comment:
Can we use an RNG object of some sort (maybe stored under `self`) instead of calling `random.randrange` directly? It's generally better not to introduce unavoidable nondeterminism like this deep inside the control flow.